lib/hyacinth/adapters/preservation_adapter/fedora3/rels_ext_properties.rb

```diff
@@ -0,0 +1,145 @@
+          apply_delta(fedora_obj, URIS::HAS_PROJECT, delta, isLiteral: true)
+          # add object->publisher string keys
+          prospective_values = publisher_string_keys_for(@hyacinth_obj)
+          delta = delta_for(fedora_obj, URIS::HAS_PUBLISHER, prospective_values)
+          apply_delta(fedora_obj, URIS::HAS_PUBLISHER, delta, isLiteral: true)
+          # add DOI URIs
+          prospective_values = dois_for(@hyacinth_obj)
+          delta = delta_for(fedora_obj, URIS::HAS_DOI, prospective_values)
+          apply_delta(fedora_obj, URIS::HAS_DOI, delta)
+          # add content model URIs
+          prospective_values = models_for(@hyacinth_obj)
+          delta = delta_for(fedora_obj, URIS::HAS_MODEL, prospective_values)
+          apply_delta(fedora_obj, URIS::HAS_MODEL, delta)
+          # add rdf type URIs
+          prospective_values = rdf_types_for(@hyacinth_obj)
+          delta = delta_for(fedora_obj, URIS::HAS_RDF_TYPE, prospective_values)
+          apply_delta(fedora_obj, URIS::HAS_RDF_TYPE, delta)
+          # add restriction flags
+          prospective_values = restrictions_for(@hyacinth_obj)
+          delta = delta_for(fedora_obj, URIS::HAS_RESTRICTION, prospective_values)
+          apply_delta(fedora_obj, URIS::HAS_RESTRICTION, delta, isLiteral: true)
+
+          return unless @hyacinth_obj.is_a? ::DigitalObject::Asset
+          # Asset-only properties
+          prospective_values = [@hyacinth_obj.resources['master'].original_filename].compact
+          delta = delta_for(fedora_obj, URIS::ORIGINAL_FILENAME, prospective_values)
+          apply_delta(fedora_obj, URIS::ORIGINAL_FILENAME, delta, isLiteral: true)
+        end
+
+        def parent_uris_for(hyacinth_obj)
+          hyacinth_obj.parent_uids.map { |uid| ::DigitalObject::Base.find(uid) }.compact.map do |parent|
+            digital_object_pids(parent)
+          end.flatten.compact
+        end
+
+        def project_string_keys_for(hyacinth_obj)
+          hyacinth_obj.projects.map(&:string_key)
+        end
+
+        def publisher_string_keys_for(hyacinth_obj)
+          hyacinth_obj.publish_entries.map { |string_key, _entry| string_key }
+        end
+
+        def dois_for(hyacinth_obj)
+          return [] unless hyacinth_obj.doi
+          [hyacinth_obj.doi].map { |x| x.start_with?("doi:") ? x : "doi:#{x}" }
+        end
+
+        def models_for(hyacinth_obj)
+          type_infos_for(hyacinth_obj, :cmodel)
+        end
+
+        def rdf_types_for(hyacinth_obj)
+          type_infos_for(hyacinth_obj, :rdf_type)
+        end
+
+        def type_infos_for(hyacinth_obj, type_key)
+          infos = []
+          return infos unless (type_info = TYPE_INFORMATION[hyacinth_obj.digital_object_type.downcase])
+          infos << type_info[type_key]
+          infos
+        end
+
+        def restrictions_for(hyacinth_obj)
+          restrictions = []
+          digital_object_data = hyacinth_obj.dynamic_field_data
```

Comment:
The term `digital_object_data` indicates top level data, while `dynamic_field_data` is a subfield of that top level data.  We should definitely change this line for that reason, since it's confusing.

Also, you won't find `restrictions` under `hyacinth_obj.dynamic_field_data`, since they're not user-defined dynamic fields.  We need to add a top level `restrictions` metadata attribute to the digital object base model.